### DISCUSSION TITLE
Curve fitting

### DISCUSSION THREAD

**Discussion by luiswirth** [NicolasKlenert/enterpolation#9] [General]
I'm sorry for this question being so broad, but I'm a little lost...

I would like to use enterpolation to do some curve fitting with NURBS.

I'm writing a pen note taking app, where I want to find a NURBS that interpolates the coordinate points of the pen as it draws a curve. Is this part of the purpose of this crate? Or is interpolation meant in a different sense here (just computing the values of a spline function?)

Is this maybe already possible with enterpolation?

I'm not even sure if I want perfect points interpolation. What I'm rather looking for, is a visual fit. I think this is referred to as geometric fitting and maybe involves solving a total least-squares problem?

What are you thoughts on this? Do you have any knowledge on this topic?

**Comment by NicolasKlenert** [comment]
I'd like the library to be able to support curve fitting in the future (after a few other algorithms to change the behaviour of the curves) but this is not implemented yet.

Curve fitting is a huge topic and I'd like to implement some basic curve fitting algorithm (like least-square optimizations) generically for every curve. However as I also would like to support multivariate curves (or surfaces to be precise) this is not an easy task.

In my opinion curve fitting with exact points are only interesting if one only has a few points to work with. If I understand correctly your users would draw curves and you want to find a good approximation of these curves. Such least-square optimizations sounds right to me. However as far as I'm aware of, there are meta-variables (like the degree of the curves you want to allow, if you allow rational splines and how many splines you want to use) which have to be set before using an algorithm to find the parameters of a good fitting curve. Least-square fitting of curves is usually done by either directly or iteratively solving some equations (done by doing matrix calculation).

It is a rather complex but rewarding task. There are many papers on the internet about curve fitting you may want to study. For startes I would recommend starting with fewer paramters (NURBS have to most of them all). Equidistand BSplines of a certain degree could be enough for what you are trying to achieve. I don't know if there are good rust librarys implementing equation solvers, but I would try to start there, as these could make your process a lot smoother.

**Comment by NicolasKlenert** [comment]
## Update to old thread:

Originally I planned to implement "Hobby splines" as these are perfect for 2D use cases (they generate rather natural looking curves, as they minimize the curvature change). However there are no correct extensions to them in 3D I could find and it is not trivial to find such an extension.

I'm writing this here mainly so that if somebody happens to find this post, a look into "Hobby splines" is worth the effort if the domain is 2D, like in the case of writing. However I do not have the plan to implement these splines anymore, as this crate shall be as general as possible. Such an algorithm better fits as a standalone crate.

I also talked about implementing surfaces. This is still an idea, however the possibility space is huge. Currently I'm leaning into noise-like functions, as these can often be arbitrarily scaled to high dimensions and only need a few informations about the elements they act upon. Fitting these is (for this crate) not achievable. The use case is just too small and the complexity to high.

## TLDR:

So my current take is: Creating and using the curves of this crate should be as easy as possible, but no assumption should be made how one should construct them. Curve fitting should therefore by its own crate, as the code is mostly orthogonal to this crate. Furthermore curve fitting often has to do with the specifics of the use case and covering all use cases seems ill-advised.
